### PR TITLE
Display correct filename with --test option

### LINKED ISSUE #39592: Cargo doctest output in nightly is incorrect
In the latest nightly, the logged output for doctests has changed to what appears to be a terser version. Unfortunately, the output is incorrect. The output shows every doctest as occurring in the same file, regardless of where the doctest is. Furthermore, doctest logging no longer includes the module/item where the doctest was declared, which, in my opinion, is a rather large usability regression.

Here is what the new, buggy output looks like:

```
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 104 ... ignored
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 116 ... ignored
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 125 ... ignored
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 13 ... ignored
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 130 ... ignored
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 14 ... ignored
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 103 ... ok
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 133 ... ok
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 112 ... ok
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 117 ... ok
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 140 ... ok
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 110 ... ok
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 101 ... ok
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 123 ... ok
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 113 ... ok
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 144 ... ok
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 120 ... ok
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 108 ... ok
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 17 ... ignored
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 101 ... ok
test /home/travis/build/SergioBenitez/Rocket/lib/src/lib.rs - line 112 ... ok
```

You can find the complete output in [the Travis logs for a recent build of Rocket](https://travis-ci.org/SergioBenitez/Rocket/builds/198938996#L676).

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #39592.

With the current files:

### test.rs

```rust
pub mod foo;

/// This is a Foo;
///
/// ```
/// println!("baaaaaar");
/// ```
#[unstable]
pub struct Foo;

/// This is a Bar;
///
/// ```
/// println!("fooooo");
/// ```
pub struct Bar;
```

### foo.rs

```rust
// note the whitespaces
/// ```
/// println!("foo");
/// ```
pub fn foo() {}
```

It displays:

```
./build/x86_64-apple-darwin/stage1/bin/rustdoc --test test.rs

running 3 tests
test test.rs - line 13 ... ok
test test.rs - line 5 ... ok
test foo.rs - line 2 ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured
```

### And with a markdown file:

```
# foo

` ``
println!("lol");
` ``

asdjnfasd

asd
```

It displays:

```
./build/x86_64-apple-darwin/stage1/bin/rustdoc --test foo.md

running 1 test
test <input> - line 3 ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured
```

r? @alexcrichton

### CODE REVIEW

**Comment by alexcrichton** [comment]
Can you be sure to add a test for this as well?

**Comment by GuillaumeGomez** [comment]
@alexcrichton: So I came up with a solution, does it seem ok to you?

File: src/tools/compiletest/src/runtest.rs [hunk 1]
```diff
@@ -1926,36 +1938,58 @@ actual:\n\
...
-                                                    .parse()
-                                                    .unwrap_or(0);
-                                   if let Ok(pos) = v.binary_search(&line) {
-                                       v.remove(pos);
-                                   } else {
-                                       self.fatal_proc_rec(
-                                           &format!("Not found doc test: \"{}\" in {:?}", s, v),
-                                           &res);
+                                   let path = tmp[0].rsplit("test ").next().unwrap();
+                                   println!("{:?}", path);
```
Comment (SergioBenitez) [inline src/tools/compiletest/src/runtest.rs]: Is this `println` really supposed to be here?

**Comment by SergioBenitez** [comment]
I don't know if this PR addresses the entire issue. Another thing that I can't do anymore is run a subset of the doctests by specifying a prefix: `cargo test some_module`. Doing this currently runs the unit tests, but not doctests, for that prefix. This is another fairly large, breaking regression. Is this fixed by this PR?

File: src/tools/compiletest/src/runtest.rs (hunk 1 above)
Comment (GuillaumeGomez) [inline src/tools/compiletest/src/runtest.rs]: No, absolutely not! Thanks a lot for notifying it!

**Comment by GuillaumeGomez** [comment]
@SergioBenitez: `cargo test` is based on name so if you want to run "specific" doc tests, you should specify the filename instead. Also, the ids of rustdoc were pretty stupid, now at least you from which file and which line it comes from (and then you can go read the corresponding code directly).

**Comment by SergioBenitez** [comment]
This is not the way things worked before, however, and frankly, I don't believe this makes any sense. There is now a strange inconsistency between `cargo test {name}` when running doctests vs. unit tests, an inconsistency that didn't exist before. I can't run all of the tests for a single module with a single command anymore. Now, I have to figure out the path to a module, run the module's unit tests, and then run the modules doctests. This is a major usability regression.

I think this is incredibly confusing. What does "the ids of rustdoc were pretty stupid" mean? It's great that I now know the file/line the test is on (by the way, are those numbers actually correct? I'm seeing some strange numbers...), but I also want to know the logical path to the module. Like I said, I have no idea which function/item is in which file/line, and forcing me to check out the file individually is a huge bummer. I want to know the file/line if I'm going to go and fix it, but I want to know the module path when I just want to know what went wrong. Both pieces of information are useful. Nothing is stupid about wanting to have the full picture.

**Comment by GuillaumeGomez** [comment]
The ids were stupid because they generally didn't give much indication. What happens if you have the same function name in multiple traits? Simply something beautiful like this:

```
func_0
func_1
func_2
func_3
func_4
...
```

Super useful, indeed. And I don't even speak when you run `rustdoc --test` on markdown files. No, clearly, this new system is way better. With the fix, all the files will be correctly printed.

**Comment by alexcrichton** [comment]
@GuillaumeGomez it seems like both indicators could be inserted in the test name, right? We could have both a small snippet of what file and also the old historical context for what's actually being tested.

**Comment by SergioBenitez** [comment]
> The ids were stupid because they generally didn't give much indication. What happens if you have the same function name in multiple traits? Simply something beautiful like this:
> 
> ```
> func_0
> func_1
> func_2
> func_3
> func_4
> ...
> ```
> 
> Super useful, indeed.

That's not what Cargo prints out at all. This is what gets printed out:

```
running 6 tests
test MyTrait::a_0 ... ok
test A::a_0 ... ok
test B::a_0 ... ok
test A::a_0 ... ok
test MyOtherTrait::a_0 ... ok
test B::a_0 ... ok
```

The logs includes the full path to the trait/function, not just the function name, thus disambiguating. I do agree that the `_n` is a bit odd, but you get used to it very quickly. I'd personally like to see `#n` instead: `MyTrait::a#0`, `MyTrait::a#1`, and so on. Or maybe even, `MyTrait::a test 1`, `MyTrait::a test 2`, and so on. Or not use it at all an disambiguate with the file/line, as I'll show below.

The other issue I have with this new format is that it uses absolute paths; they're just too long. I don't care about the absolute path to my project; I ran `cargo test`, I'm in the workspace. I think the path should be relative to the current Cargo workspace or Cargo root.

I think something like this would look nice:

```
test MyTrait::a (src/lib.rs:2) ... ok
test A::a (src/lib.rs:9) ... ok
test B::a (src/lib.rs:33) ... ok
test A::a (src/lib.rs:19) ... ok
test MyOtherTrait::a (src/lib.rs:40) ... ok
test B::a (src/lib.rs:26) ... ok
```

This is concise, informative, and includes all of the old and new information.

**Comment by GuillaumeGomez** [comment]
Having the two pieces of information could be useful, indeed. I can add it back (without the `_n` if you don't mind). Could you open an issue for it please?

**Pull request merged**